File: diagnostic_management/diagnostic_management/report/sales_register_with_doctor/sales_register_with_doctor.js
// Sales Register with Doctor — ERPNext's Sales Register filters, plus a
// Doctor filter fed from the distinct `custom_doctor` values actually recorded
// on submitted invoices (it's a free-text field, so there's no master to link).
frappe.query_reports["Sales Register with Doctor"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			width: "80",
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "doctor",
			label: __("Doctor"),
			fieldtype: "Autocomplete",
			get_data: function (txt) {
				return new Promise((resolve) => {
					frappe.call({
						method: "diagnostic_management.diagnostic_management.report.sales_register_with_doctor.sales_register_with_doctor.doctor_options",
						args: {
							txt: txt,
							page_len: 50,
							filters: {
								company: frappe.query_report.get_filter_value("company"),
							},
						},
						callback: function (r) {
							resolve((r.message || []).map((row) => row[0]));
						},
					});
				});
			},
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "customer_group",
			label: __("Customer Group"),
			fieldtype: "Link",
			options: "Customer Group",
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "mode_of_payment",
			label: __("Mode of Payment"),
			fieldtype: "Link",
			options: "Mode of Payment",
		},
		{
			fieldname: "owner",
			label: __("Owner"),
			fieldtype: "Link",
			options: "User",
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
		},
		{
			fieldname: "brand",
			label: __("Brand"),
			fieldtype: "Link",
			options: "Brand",
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
		},
		{
			fieldname: "include_payments",
			label: __("Show Ledger View"),
			fieldtype: "Check",
			default: 0,
		},
	],
};

// Same accounting-dimension filters the stock Sales Register exposes.
if (typeof erpnext !== "undefined" && erpnext.utils && erpnext.utils.add_dimensions) {
	erpnext.utils.add_dimensions("Sales Register with Doctor", 7);
}

File: diagnostic_management/diagnostic_management/report/sales_register_with_doctor/sales_register_with_doctor.py
"""Sales Register with the referring Doctor.

ERPNext's Sales Register with one extra column: `Sales Invoice.custom_doctor`,
the referring doctor picked at Billing. Rather than fork the report (and inherit
the maintenance of every column, tax breakdown and dimension ERPNext adds later)
this calls ERPNext's own `_execute()` through its supported extension point,
`additional_table_columns` — the same hook `item_wise_sales_register` uses.

That means every column, filter and total stays identical to the stock report
and keeps tracking upstream; the Doctor column is selected in the same query,
so there is no per-row lookup.

The Doctor filter is applied after the fact because ERPNext's `get_conditions()`
only understands its own filter set. Rows come back as dicts keyed by fieldname,
so filtering on `custom_doctor` is exact.
"""

import frappe
from frappe import _

from erpnext.accounts.report.sales_register.sales_register import _execute

DOCTOR_FIELD = "custom_doctor"


def _doctor_column() -> dict:
	return {
		"label": _("Doctor"),
		"fieldname": DOCTOR_FIELD,
		"fieldtype": "Data",
		"width": 160,
	}


def execute(filters=None):
	filters = frappe._dict(filters or {})

	# Pulled out before delegating — ERPNext would reject an unknown filter.
	doctor = (filters.pop("doctor", None) or "").strip()

	if not frappe.db.has_column("Sales Invoice", DOCTOR_FIELD):
		# Custom field not installed yet (fresh site, pre-migrate). Degrade to
		# the stock report rather than throwing a SQL error at the user.
		frappe.msgprint(
			_("Sales Invoice has no {0} field yet — showing the standard Sales Register. "
			  "Run a migrate to install it.").format(frappe.bold(DOCTOR_FIELD)),
			indicator="orange", alert=True,
		)
		return _execute(filters)

	result = _execute(filters, additional_table_columns=[_doctor_column()])

	# _execute returns a variable-length tuple across ERPNext versions
	# (columns, data[, message, chart, report_summary]). Keep whatever else it
	# hands back instead of assuming a 2-tuple.
	columns, data = result[0], result[1]
	rest = tuple(result[2:])

	if doctor:
		wanted = doctor.casefold()
		data = [
			row for row in data
			if isinstance(row, dict) and (row.get(DOCTOR_FIELD) or "").strip().casefold() == wanted
		]

	return (columns, data) + rest


@frappe.whitelist()
def doctor_options(doctype=None, txt=None, searchfield=None, start=0, page_len=20, filters=None):
	"""Link-style options for the Doctor filter.

	`custom_doctor` is a free-text Data field, so the choices are the distinct
	values actually recorded on invoices — no separate master to point a Link
	filter at, and no stale names.
	"""
	conditions, values = ["COALESCE(custom_doctor, '') != ''"], []
	if txt:
		conditions.append("custom_doctor LIKE %s")
		values.append(f"%{txt}%")
	company = (filters or {}).get("company") if isinstance(filters, dict) else None
	if company:
		conditions.append("company = %s")
		values.append(company)

	return frappe.db.sql(
		f"""
		SELECT DISTINCT custom_doctor
		FROM `tabSales Invoice`
		WHERE docstatus = 1 AND {" AND ".join(conditions)}
		ORDER BY custom_doctor ASC
		LIMIT %s, %s
		""",
		(*values, int(start), int(page_len)),
	)
